### PR TITLE
ATMOS: Lavaland atmospheric vents, caps, and even more regrettable balance decisions! [WIP]

### PR DESCRIPTION
**Pull request by Buggy123**
Adds a more balanced and server-friendly and involved counterpart to Gas Miners as suggested by @ninjanomnom, as a method for getting non-finite supplies of gas for atmospheric projects.

The stuff

**Atmospheric Vents**
Atmospheric vents spawn on lavaland, one for each type of gas found in the randomized atmosphere.

**Vent Caps** (non-lame name pending)
You stick these on a Atmospheric Vent to gather gas and direct it into pipes.

By default, it requires no power and produces a decent amount of gas. If you need more, you can supply (quite a bit) of power to widen the vent, increasing production. Note that the gas output is scaled to the quantity of the gas in the atmosphere, so gathering anything with a low concentration is substantially slower.

Maintaining the new rate of outflow doesn't require power, but only if you can keep up with it! Too much pressure in the vent will clog it and slow it back down. And beware, you **really** don't want to damage the machinery which is actively containing a immense outflow of very high pressure gas; keep those fauna _away_ from the vent!<br/>
<br/>
 
As far as balance, these don't actually add much of anything that you can't do already; you can already use Lavaland's atmospheric turfs to gather infinite quantities of any gas in the atmosphere via scrubbers.

The problem with this method, however, is that it breaks the game; several dozen+ scrubbers will lag and break the air alarm GUI, in extreme cases outright freezing your client and forcing you to restart. It also lags the server rather noticably. This method does the same thing and requires similar amounts of setup time, but in a less broken way with less server impact.<br/>
<br/>

Note: EXTREMELY WIP. Literally only putting it out this early because it's such a perfect counterpart to #49067. Todo:

- [ ] Finish and refine the code.
- [ ] Add the finishing touches.
- [ ] Finish the UI.
- [ ] Balance it a bit.
- [ ] Add animated codersprites to gas vent.
- [ ] Maybe get non-codersprites too?
- [ ] Figure out less-lame alternative names for 'gas vent' and 'growing'.
<br/>
<br/>
:cl:
Add: The Nanotrasen Office of Lavaland Exploration and Gathering has discovered the source of Lavaland's strange atmosphere; a series of vents which have incorporated the naturally-occuring bluespace minerals into their structure, producing a infinite supply of gas! NOLAG has begun distributing plans for vent capping machinery to aid in harvesting useful gases directly from these vents, in hopes of reducing the local space-time pollution produced by the current practice of using ill-designed contaminant scrubbing equipment to harvest gas from the atmosphere.

/:cl:

### CODE REVIEW

**Comment by TheVekter** [comment]
This is a neat idea and a decent alternative to gas miners. I like it.

**Comment by HomocidlePancakes** [comment]
Ok. So how are you going to get the gas back to station? In canisters? If so, then this is a bruh moment.

**Comment by Buggy123** [comment]
>Ok. So how are you going to get the gas back to station? In canisters? If so, then this is a bruh moment.

Shuttles can be used as impromptu gas tanks or have storage pipe networks placed in them, which _should_ automatically connect whenever it moves but that's kinda fucky still.

Failing that, canisters with supercooled gas can hold up to about 360,000 moles of gas. Each.

**Comment by Tlaltecuhtli** [comment]
why does the gayser look like a hologram

**Comment by ninjanomnom** [comment]
You don't need bluespace to explain a gas vent

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

File: code/modules/atmospherics/machinery/components/unary_devices/ventcap.dm [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+
+	var/datum/looping_sound/ventcap/soundloop
+
+
+/obj/machinery/portable_atmospherics/ventcap/RefreshParts()
+	var/list/pinfo = list()
+		
+	for(var/obj/item/stock_parts/matter_bin/M in component_parts)
+		pinfo[1]++
+		pinfo["[pinfo[1]]"] += M.rating
```
[the remark this answers is not in the document]
Comment (Buggy123) [reply code/modules/atmospherics/machinery/components/unary_devices/ventcap.dm]: Yeah, I thought having it be part-count independent would be good but it turned into nuclear-grade shitcode.

File: code/datums/looping_sounds/machinery_sounds.dm [hunk 2]
```diff
@@ -61,3 +61,18 @@
...
+/datum/looping_sound/ventcap
+	mid_sounds = list(
+		'sound/weather/ashstorm/inside/active_mid1.ogg'=1,
+		'sound/weather/ashstorm/inside/active_mid2.ogg'=1,
+		'sound/weather/ashstorm/inside/active_mid3.ogg'=1
+		)
+	mid_length = 80
+	start_sound = 'sound/weather/ashstorm/inside/active_start.ogg'
+	start_length = 130
+	end_sound = 'sound/weather/ashstorm/inside/active_end.ogg'
```
[the remark this answers is not in the document]
Comment (Buggy123) [reply code/datums/looping_sounds/machinery_sounds.dm]: I checked thoroughly, there is no "gas loudly flowing through pipes and machinery" noise in the sound files. Ash storms sounds are the closest.

**Comment by Buggy123** [comment]
>why does the gayser look like a hologram

Codersprites

File: code/datums/looping_sounds/machinery_sounds.dm (hunk 2 above)
[the remark this answers is not in the document]
Comment (ninjanomnom) [reply code/datums/looping_sounds/machinery_sounds.dm]: Then add a new one

**Comment by PKPenguin321** [comment]
>Atmospheric vents spawn on lavaland, one for each type of gas found in the randomized atmosphere.

More than one would probably be okay. Maybe in the far future a way to bore a hole to make your own one.

**Review by stylemistake — DISMISSED** [review]
TGUI part is ok

**Comment by Buggy123** [comment]
>TGUI part is ok


TGUI part isn't done. Nor is the atmos code, or the non-atmos code. I DID recently finish the animated icon_states and now I'm just slowly chipping away at making sure it animates properly without any way to skip between states.

When I said "EXTREMELY WIP", I really mean it. Assume at least half of the code will be completely refactored and/or added to.

File: code/game/objects/structures/lavaland/vent.dm [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+	for(var/gas in env.gases)
+		if(env.gases[gas][MOLES] > 0)
+			possible_gases += gas
+
+	for(var/obj/machinery/atmospherics/components/unary/portables_connector/atmosphere_vent/vent in GLOB.atmospheric_vents)
+		if(vent.z == src.z)
+			possible_gases -= vent.gastype
+
+	if(!possible_gases.len && naturally_spawned)
+		qdel(src)
```
[the remark this answers is not in the document]
Comment (Buggy123) [reply code/game/objects/structures/lavaland/vent.dm]: I guess, I just don't want to spam lavaland with random 3x3 excavated squares. Which also might require, additionally, a moderately sized area around them which blocks ruin formation. Too much stuff on lavaland and it starts interfering with ruin and boss spawn, and I _don't_ want to do that.

[the remark this answers is not in the document]
Comment (ninjanomnom) [reply code/game/objects/structures/lavaland/vent.dm]: You already restrict to a max of 7 or something though, that's prety reasonable and if the atmosphere has a low amount of gases more of each type seems fine

File: code/game/objects/structures/lavaland/vent.dm [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+	var/list/possible_gases = list()
+	
+	for(var/gastype in GLOB.meta_gas_info)
+		env.assert_gas(gastype)
+	for(var/gas in env.gases)
+		if(env.gases[gas][MOLES] > 0)
+			possible_gases += gas
+
+	for(var/obj/machinery/atmospherics/components/unary/portables_connector/atmosphere_vent/vent in GLOB.atmospheric_vents)
+		if(vent.z == src.z)
```
[the remark this answers is not in the document]
Comment (Buggy123) [reply code/game/objects/structures/lavaland/vent.dm]: should be done

File: code/game/objects/structures/lavaland/vent.dm [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+		return
+
+	var/datum/gas_mixture/env = T.return_air()
+	var/list/possible_gases = list()
+	
+	for(var/gastype in GLOB.meta_gas_info)
+		env.assert_gas(gastype)
+	for(var/gas in env.gases)
+		if(env.gases[gas][MOLES] > 0)
+			possible_gases += gas
```
[the remark this answers is not in the document]
Comment (Buggy123) [reply code/game/objects/structures/lavaland/vent.dm]: done

File: code/game/objects/structures/lavaland/vent.dm [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+	for(var/obj/machinery/atmospherics/components/unary/portables_connector/atmosphere_vent/vent in GLOB.atmospheric_vents)
```
[the remark this answers is not in the document]
Comment (Buggy123) [reply code/game/objects/structures/lavaland/vent.dm]: done, though I don't think it saves much effort

File: code/game/objects/structures/lavaland/vent.dm [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+	can_unwrench = FALSE
+	pixel_x = -16
+	resistance_flags = INDESTRUCTIBLE | LAVA_PROOF | FIRE_PROOF | UNACIDABLE | ACID_PROOF
+	
+	var/naturally_spawned = FALSE // Naturally spawned vents have this set to TRUE. This is so these can be manually spawned without having the vent immediately qdel itself if other vents already exist.
+	var/gastype = null
+
+/obj/machinery/atmospherics/components/unary/portables_connector/atmosphere_vent/Initialize()
+	. = ..()
+	addtimer(CALLBACK(src, .proc/set_gas), 10) // This breaks if done immediately for whatever reason.
```
[the remark this answers is not in the document]
Comment (Buggy123) [reply code/game/objects/structures/lavaland/vent.dm]: done

File: code/game/objects/structures/lavaland/vent.dm [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+	if(!possible_gases.len && naturally_spawned)
+		qdel(src)
+		return
+
+	GLOB.atmospheric_vents += src
+
+	gastype = pick(possible_gases)
+
+/obj/machinery/atmospherics/components/unary/portables_connector/atmosphere_vent/Destroy()
+	. = ..()
```
[the remark this answers is not in the document]
Comment (Buggy123) [reply code/game/objects/structures/lavaland/vent.dm]: done

**Comment by optimumtact** [comment]
how about you come back when this is actually done

**Comment by Buggy123** [comment]
:(

**Pull request closed without merging**